overops-functions/src/main/java/com/takipi/udf/microsoftteams/MicrosoftTeamsChannelFunction.java

```diff
@@ -0,0 +1,195 @@
+import com.takipi.api.client.data.service.SummarizedService;
+import com.takipi.api.client.data.view.SummarizedView;
+import com.takipi.api.client.request.event.EventRequest;
+import com.takipi.api.client.request.event.EventsRequest;
+import com.takipi.api.client.result.event.EventResult;
+import com.takipi.api.client.result.event.EventsResult;
+import com.takipi.api.client.util.event.EventUtil;
+import com.takipi.api.client.util.validation.ValidationUtil;
+import com.takipi.api.client.util.view.ViewUtil;
+import com.takipi.api.core.url.UrlClient;
+import com.takipi.common.util.CollectionUtil;
+import com.takipi.udf.ContextArgs;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.ISODateTimeFormat;
+
+import java.util.Base64;
+import java.util.List;
+
+import static com.takipi.udf.microsoftteams.MicrosoftTeamsUtil.*;
+
+public class MicrosoftTeamsChannelFunction {
+
+    public static final int MINUTES_TIME_SPAN = 5;
+
+    public static String validateInput(String rawInput) {
+        return getInput(rawInput).toString();
+    }
+
+    private static MicrosoftTeamsInput getInput(String rawInput) {
+        if (Strings.isNullOrEmpty(rawInput))
+            throw new IllegalArgumentException("Input is empty");
+
+        MicrosoftTeamsInput input;
+
+        try {
+            input = MicrosoftTeamsInput.of(rawInput);
+        } catch (Exception e) {
+            throw new IllegalArgumentException(e.getMessage(), e);
+        }
+
+        input.checkUrl();
+
+        return input;
+    }
+
+    public static void execute(String rawContextArgs, String rawInput) {
+        try {
+            executeImplementation(rawContextArgs, rawInput);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    private static void executeImplementation(String rawContextArgs, String rawInput) {
+        logUDFInput(rawContextArgs, rawInput);
+        MicrosoftTeamsInput input = getInput(rawInput);
+        ContextArgs args = (new Gson()).fromJson(rawContextArgs, ContextArgs.class);
+
+        if (!args.validate())
+            throw new IllegalArgumentException("Bad context args: " + rawContextArgs);
+        if (!args.eventValidate())
+            return;
+
+        ApiClient apiClient = args.apiClient();
+        UrlClient.Response<EventResult> eventResultResponse = getEventResultResponse(args, apiClient);
+        System.out.println("Got exceptionClassName, exception type, exceptionLocationPath, stack_frames, deployment");
+        MicrosoftTeamsUtil.EventData eventData = getEventData(apiClient, args);
+        System.out.println("Got applications, environments, deployments and servers");
+        MicrosoftTeamsUtil.TimeSlot timeSlot = getTimeSlot(MINUTES_TIME_SPAN);
+
+        SummarizedView view = getSummarizedView(rawContextArgs, args, apiClient);
+        System.out.println("Got viewId");
+
+        Graph eventsGraph = ViewUtil.getEventsGraph(apiClient, args.serviceId, view.id,
```

Comment:
What is this graph used for?
Getting a graph is one of the most expensive operations possible, and it seems this is only used to get the server/application/deployment?
For what purpose?

Reply:
I'll refine my Q -
What are the machine/application/deployment names *should* be?
The ones relating to the first snapshot ever for this event?
The ones relating to all the places this event happened so far?

Reply:
Here screenshots where related data I could find in the backend
![image](https://user-images.githubusercontent.com/1032313/67384014-fffb3c80-f598-11e9-84fc-ceaae268015e.png)
![image](https://user-images.githubusercontent.com/1032313/67384203-64b69700-f599-11e9-898c-741b688368b3.png)
![image](https://user-images.githubusercontent.com/1032313/67384501-ec9ca100-f599-11e9-950d-67316db403f8.png)


In Microsoft Teams they will be reflected
![image](https://user-images.githubusercontent.com/1032313/67384679-4d2bde00-f59a-11e9-9b9c-d7a8613c3520.png)

Reply:
OK
Starting with api-client 2.29.0, you'll have the following fields on `EventResult` -
`introduced_by_server` - the server in which this event first happened.
`introduced_by_application` - the application in which this event first happened.

The matching field for deployments is called `introduced_by`, and it's been there from earlier versions.

Will this solve your issue?

NOTE - while the fields themselves exist, the backend will only start populating them in a few days.
I just wanted to make sure that you can easily use a version that will allow you to work :)

Reply:
@dor-levi thanks, I will apply when 2.29.0 version will be available on prod

Reply:
@dor-levi  done